src/Microsoft.OData.Core/BindingPathHelper.cs

```diff
@@ -23,15 +23,16 @@ internal static class BindingPathHelper
         /// <returns>True if the path of navigation property in current scope is matching the <paramref name="bindingPath"/>.</returns>
         public static bool MatchBindingPath(IEdmPathExpression bindingPath, IList<ODataPathSegment> parsedSegments)
         {
-            List<string> paths = bindingPath.PathSegments.ToList();

```

Comment:
I would question, in the code path below, whether IEnumerable is more efficient than calling ToList() since we are looping through the results multiple times.  In particular, line 27 will loop through the enumerable to count the number of segments and then lines 48/60 will call ElementAt() for each property/type segment in the path